Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Eel/ElasticSearchQueryBuilder.php

```diff
@@ -561,64 +555,27 @@ public function fetch() {
-		}
-
-		$this->totalItems = $hits['total'];
-
-		if ($hits['total'] === 0) {
-			return array();
-		}
 
-		$nodes = array();
-		$elasticSearchHitPerNode = array();
-
-		/**
-		 * TODO: This code below is not fully correct yet:
-		 *
-		 * We always fetch $limit * (numerOfWorkspaces) records; so that we find a node:
-		 * - *once* if it is only in live workspace and matches the query
-		 * - *once* if it is only in user workspace and matches the query
-		 * - *twice* if it is in both workspaces and matches the query *both times*. In this case we filter the duplicate record.
-		 * - *once* if it is in the live workspace and has been DELETED in the user workspace (STILL WRONG)
-		 * - *once* if it is in the live workspace and has been MODIFIED to NOT MATCH THE QUERY ANYMORE in user workspace (STILL WRONG)
-		 *
-		 * If we want to fix this cleanly, we'd need to do an *additional query* in order to filter all nodes from a non-user workspace
-		 * which *do exist in the user workspace but do NOT match the current query*. This has to be done somehow "recursively"; and later
-		 * we might be able to use https://github.com/elasticsearch/elasticsearch/issues/3300 as soon as it is merged.
-		 */
-		foreach ($hits['hits'] as $hit) {
-			// with ElasticSearch 1.0 fields always returns an array,
-			// see https://github.com/Flowpack/Flowpack.ElasticSearch.ContentRepositoryAdaptor/issues/17
-			if (is_array($hit['fields']['__path'])) {
-				$nodePath = current($hit['fields']['__path']);
-			} else {
-				$nodePath = $hit['fields']['__path'];
-			}
-			$node = $this->contextNode->getNode($nodePath);
-			if ($node instanceof NodeInterface) {
-				$nodes[$node->getIdentifier()] = $node;
-				$elasticSearchHitPerNode[$node->getIdentifier()] = $hit;
-				if ($this->limit > 0 && count($nodes) >= $this->limit) {
-					break;
-				}
-			}
+		$hits = array();
+		$this->result['nodes'] = array();
+		if(array_key_exists('hits', $treatedContent) && is_array($treatedContent['hits']) && count($treatedContent['hits']) > 0) {
+			$hits = $treatedContent['hits'];
+			$this->result['nodes'] = $this->convertHitsToNodes($treatedContent['hits']);
 		}
 
 		if ($this->logThisQuery === TRUE) {
-			$this->logger->log('Query Log (' . $this->logMessage . ') Number of returned results: ' . count($nodes), LOG_DEBUG);
+			$this->logger->log(sprintf('Query Log (%s): %s -- execution time: %s ms -- Limit: %s -- Number of results returned: %s -- Total Results: %s',
+				$this->logMessage, json_encode($this->request), (($timeAfterwards - $timeBefore) * 1000), $this->limit, count($hits['hits']), $hits['total'])
+				, LOG_DEBUG);
 		}
 
-		$this->elasticSearchHitsIndexedByNodeFromLastRequest = $elasticSearchHitPerNode;
+		$this->result['totalItemCount'] = $hits['total'];
```

Comment:
any reason why we convert the result to a different format and not just say $this->result = $hits?